### PR TITLE
Fetch user data based on name

### PR DESCRIPTION
**Pull request by IvanJanuary**

### CODE REVIEW

File: PersonDataRequest/PersonDataRequest/Controller/ViewController.swift [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let selectedName = nameArray[indexPath.row]
+        let storyboard = UIStoryboard(name: "Main", bundle: Bundle.main)
+        
+        guard let destination = storyboard.instantiateViewController(withIdentifier: "DataViewController") as? DataViewController else {
+            return
+        }
+        destination.selectedName = selectedName
+        navigationController?.pushViewController(destination, animated: true)
+                print("\(indexPath.row)")
```
Comment (Lana-Resquit) [inline PersonDataRequest/PersonDataRequest/Controller/ViewController.swift:56]: This print is not needed anymore

File: PersonDataRequest/PersonDataRequest/Controller/DataViewController.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        nameLabel.text = selectedName
+        
+        netRequest()
+    }
+    
+    func netRequest() {
```
Comment (Lana-Resquit) [inline PersonDataRequest/PersonDataRequest/Controller/DataViewController.swift:27]: This should be renamed to something like `queryUserData` or `fetchData`

File: PersonDataRequest/PersonDataRequest/Controller/DataViewController.swift [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        guard let selectedName = selectedName else { return }
+        let genderUrl = "https://api.genderize.io/?name=\(selectedName)"
+        let ageUrl = "https://api.agify.io/?name=\(selectedName)"
+        
+        let apiHelper = ApiHelper()
+        
+        apiHelper.makeRequest(urlString: genderUrl, t: GenderData.self) { genderData in
```
Comment (Lana-Resquit) [inline PersonDataRequest/PersonDataRequest/Controller/DataViewController.swift:34]: You need to add `[weak self]` before `genderData`, this is a rule for closures where `self` is used

File: PersonDataRequest/PersonDataRequest/Controller/DataViewController.swift [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            guard let genderData = genderData else { return }
+            DispatchQueue.main.async {
+                self.genderUpdateUI(with: genderData)
+            }
+        }
+        
+        apiHelper.makeRequest(urlString: ageUrl, t: AgeData.self) { ageData in
```
Comment (Lana-Resquit) [inline PersonDataRequest/PersonDataRequest/Controller/DataViewController.swift:41]: And here as well

File: PersonDataRequest/PersonDataRequest/Controller/ApiHelper.swift [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            guard let data = data else {
+                completion(nil)
+                return }
+            
+            
+            do {
+                let decodedData = try JSONDecoder().decode(T.self, from: data)
+                completion(decodedData)
+            } catch {
+                print("Decoding error JSON age: \(error)")
```
Comment (Lana-Resquit) [inline PersonDataRequest/PersonDataRequest/Controller/ApiHelper.swift:33]: We need to call a `completion(nil)` if the error comes

File: PersonDataRequest/PersonDataRequest/Controller/ApiHelper.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    func makeRequest<T: Decodable>(urlString: String, t: T.Type, completion: @escaping (T?) -> Void) {
+        guard let url = URL(string: urlString) else {
+            completion(nil)
+            return
+        }
+        
+        let task = URLSession.shared.dataTask(with: url) { (data, response, error) in
+            if let error = error {
+                print("Gender request error: \(error.localizedDescription)")
+                return
```
Comment (Lana-Resquit) [inline PersonDataRequest/PersonDataRequest/Controller/ApiHelper.swift:21]: You need to remove this return or call the `completion(nil`) before return

**Review by Lana-Resquit — CHANGES_REQUESTED** [review]

File: PersonDataRequest/PersonDataRequest/Controller/ViewController.swift (hunk 1 above)
Comment (IvanJanuary) [reply PersonDataRequest/PersonDataRequest/Controller/ViewController.swift:56]: Removed

File: PersonDataRequest/PersonDataRequest/Controller/DataViewController.swift (hunk 2 above)
Comment (IvanJanuary) [reply PersonDataRequest/PersonDataRequest/Controller/DataViewController.swift:27]: Ok. Done

File: PersonDataRequest/PersonDataRequest/Controller/DataViewController.swift (hunk 3 above)
Comment (IvanJanuary) [reply PersonDataRequest/PersonDataRequest/Controller/DataViewController.swift:34]: I added [weak self] before genderData and made self? in the next row

File: PersonDataRequest/PersonDataRequest/Controller/DataViewController.swift (hunk 4 above)
Comment (IvanJanuary) [reply PersonDataRequest/PersonDataRequest/Controller/DataViewController.swift:41]: I did the same thing here

File: PersonDataRequest/PersonDataRequest/Controller/ApiHelper.swift (hunk 5 above)
Comment (IvanJanuary) [reply PersonDataRequest/PersonDataRequest/Controller/ApiHelper.swift:33]: Done

File: PersonDataRequest/PersonDataRequest/Controller/ApiHelper.swift (hunk 6 above)
Comment (IvanJanuary) [reply PersonDataRequest/PersonDataRequest/Controller/ApiHelper.swift:21]: I removed it

File: PersonDataRequest/PersonDataRequest/Controller/ApiHelper.swift [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    func makeRequest<T: Decodable>(urlString: String, t: T.Type, completion: @escaping (T?) -> Void) {
+        guard let url = URL(string: urlString) else {
+            completion(nil)
+            return
+        }
+        
+        let task = URLSession.shared.dataTask(with: url) { (data, response, error) in
+            if let error = error {
+                print("Gender request error: \(error.localizedDescription)")
+                completion(nil)
```
Comment (Lana-Resquit) [inline PersonDataRequest/PersonDataRequest/Controller/ApiHelper.swift:21]: It is the wrong one, you needed to remove the return OR add completion(nil)
So it should be:
```
if let error = error {
                print("Gender request error: \(error.localizedDescription)")
                completion(nil)
                return
 }
```
OR

```
if let error = error {
                print("Gender request error: \(error.localizedDescription)")
}
```

**Review by Lana-Resquit — CHANGES_REQUESTED** [review]

**Review by Lana-Resquit — APPROVED** [review]

**Pull request merged**